openvdb/points/AttributeArray.h

```diff
@@ -836,6 +1019,92 @@ UnitVecCodec::encode(const math::Vec3<T>& val, StorageType& data)
+    bool rangeChecking/* = false*/)
+{
+    // ensure both arrays have float-float or integer-integer value types
+    assert(sourceArray.valueTypeIsFloatingPoint() != this->valueTypeIsFloatingPoint());
+    // ensure both arrays have been loaded from disk (if delay-loaded)
+    assert(sourceArray.isDataLoaded() && this->isDataLoaded());
+    // ensure storage size * stride matches on both arrays
+    assert(this->storageTypeSize()*this->stride() == sourceArray.storageTypeSize()*sourceArray.stride());
+
+    const size_t bytes(sourceArray.storageTypeSize()*sourceArray.stride());
+    const char* const sourceBuffer = sourceArray.dataAsByteArray();
+    char* const targetBuffer = this->dataAsByteArray();
+    assert(sourceBuffer && targetBuffer);
+
+    if (rangeChecking && this->isUniform()) {
+        OPENVDB_THROW(IndexError, "Cannot copy array data as target array is uniform.");
+    }
+
+    const bool sourceIsUniform = sourceArray.isUniform();
+
+    const Index sourceDataSize = rangeChecking ? sourceArray.dataSize() : 0;
+    const Index targetDataSize = rangeChecking ? this->dataSize() : 0;
+
+    for (IterT it(iter); it; ++it) {
+        const Index sourceIndex = sourceIsUniform ? 0 : it.sourceIndex();
+        const Index targetIndex = it.targetIndex();
+
+        if (rangeChecking) {
+            if (sourceIndex >= sourceDataSize) {
+                OPENVDB_THROW(IndexError,
+                    "Cannot copy array data as source index exceeds size of source array.");
+            }
+            if (targetIndex >= targetDataSize) {
+                OPENVDB_THROW(IndexError,
+                    "Cannot copy array data as target index exceeds size of target array.");
+            }
+        } else {
+            // range-checking asserts
+            assert(sourceIndex < sourceArray.dataSize());
+            assert(targetIndex < this->dataSize());
+        }
+
+        const size_t targetOffset(targetIndex * bytes);
+        const size_t sourceOffset(sourceIndex * bytes);
+
+        std::memcpy(targetBuffer + targetOffset, sourceBuffer + sourceOffset, bytes);
+    }
+}
+
+template <typename IterT>
+void AttributeArray::copyValues(const AttributeArray& sourceArray, const IterT& iter,
+    bool preserveUniformity/* = true*/)
+{
+    const Index bytes = sourceArray.storageTypeSize();
+    if (bytes != this->storageTypeSize()) {
+        OPENVDB_THROW(TypeError, "Cannot copy array data due to mis-match in storage type sizes.");
+    }
+
+    // ensure both arrays have been loaded from disk
+    sourceArray.loadData();
+    this->loadData();
+
+    const bool uniform = this->isUniform();
+
+    // if the target array is uniform, expand it first
+    if (uniform) {
+        this->expand();
+    }
+
+    this->copyValuesUnsafe(sourceArray, iter, /*rangeChecking = */true);
+
+    // attempt to compact target array if previously uniform
+    if (preserveUniformity && uniform) {
+        this->compact();
+    }
```

Comment:
How about simply a compact argument instead of preserveUniformity which tries to do this regardless. You lose the automatic ability to compact if originally uniform, but you just pass ```target.isUniform()``` as the argument instead

Reply:
I've changed this to match your suggestion.